### PR TITLE
Bring Baker Franke's text compression widget in as an External HTML level

### PR DESCRIPTION
**Pull request by joshlory**
This is an import of Baker Franke's text compression widget. The widget reuses the "External HTML" DSL type which was added in https://github.com/code-dot-org/code-dot-org/pull/1271.

text-compression

All credit to Baker for the widget. You can view his original widget here: http://cspdashboard-c9-bfranke.c9.io/TextCompression/TextCompressionTool.html

### CODE REVIEW

File: dashboard/public/text-compression/text-compression.html [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+          dialog.hide();
+        });
+      } else {
+        selectedPoem = poemSelect.selectedIndex;
+        setupPoemWithText(poemsList[selectedPoem]);
+      }
+      compress();
+    }
+
+    function addOption(text) {
```
Comment (islemaster) [inline dashboard/public/text-compression/text-compression.html]: More descriptive `addOptionToPoemSelect`?

File: dashboard/app/assets/javascripts/levels/external.js [hunk 2]
```diff
@@ -0,0 +1,2 @@
+//= require codemirror
+//= require_self
```
Comment (wjordan) [inline dashboard/app/assets/javascripts/levels/external.js]: I don't think the require_self is strictly needed unless there's javascript within this file that you want to included into the bundle. just fyi, there's no harm in leaving it as-is

File: dashboard/public/text-compression/text-compression.html [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+      compression /= 100;
+      document.getElementById("data").innerHTML = "      poem size: " + poemSize + " bytes\n";
+      document.getElementById("data").innerHTML += "dictionary size: " + dictSize + " bytes\n";
+      document.getElementById("data").innerHTML += "          total: " + total + " bytes\n";
+      document.getElementById("data").innerHTML += "    compression: " + compression + "%\n";
+    }
+
+    function escapeRegExp(str) {
+      return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+    }
```
Comment (islemaster) [inline dashboard/public/text-compression/text-compression.html]: This could really use a comment.

**Comment by wjordan** [comment]
just glanced at the coremirror integration, looks straightforward enough. Glad the existing codemirror library could be included so easily!

**Comment by islemaster** [comment]
So cool. :+1:

**Pull request merged**